### ISSUE TITLE
Observability and metrics - Save to Redis

### ISSUE THREAD

**Issue by chrismeyersfsu** [ansible/awx#9012]
TL;DR Store metrics in Redis so they are accessible to the API.

Metrics live in various process spaces but we want to display/view them via other process memory spaces. For example, metrics about the task manager come from a dispatcher process that runs task manager code. The dispatcher processes space may be short lived. We want to present or view, in the API, the metrics that occur in the task manager, but the API is a separate process. We need some IPC mechanism to accomplish this. We can use Redis.

Another consumer of the metrics data would be a periodic task that outputs the values to a file on disk. This file can be collected by sosreport for offline reconstruction of past performance.

Yet another consumer of this data could be a command line tool to aid in debugging performance live.

### Requirements
1. Metric data stored in Redis should be convertible to Prometheus data types: Counter, Gauge, Histogram, Summary https://prometheus.io/docs/concepts/metric_types/ . We should serialize the metrics into JSON (look for a fast JSON serialize/serialize lib)

2. Metrics should be stored efficiently. Maybe metrics about metrics overhead.

We can use the Redis hash data type to store key -> value metrics https://redis.io/commands#hash possible organization:

Single Hash
```
HMSET awx_metrics task_manager_pending_total 2
HMSET awx_metrics task_manager_running_total 5
HMSET awx_metrics api_jobs_created 100
HMSET awx_metrics api_workflow_jobs_created 5
HMSET awx_metrics callback_receiver_events_processed 10000
```

Multiple Hashes
```
HMSET awx_metrics_task_manager pending_total 2
HMSET awx_metrics_task_manager running_total 5
HMSET awx_metrics_api jobs_created 100
HMSET awx_metrics_api workflow_jobs_created 5
HMSET awx_metrics_callback_receiver events_processed 10000
```

#### Notes
There will be many cases were many metrics need to be updated for a subsystem per "loop". It can be costly to perform multiple redis operations per-loop if we are in a critical section (i.e. the callback receiver). We could try to use redis async but this can complicate the code. Instead, we should use redis pipeline feature. You can think of the pipeline feature as batching or a transaction.

**Comment by unlikelyzero** [comment]
@kdelee has some testing notes and will add here

**Comment by kdelee** [comment]
we will end up testing this once it is served to the metrics endpoint https://github.com/ansible/awx/issues/9019#issuecomment-771926550

**Comment by ryanpetrello** [comment]
Closing, cc @kdelee @fosterseth
